File: commands/help.pl
#!/usr/bin/perl

use lib 'src';
use Helper;
use Henzell::Config qw/%CMDPATH/;
use LearnDB;

Helper::help("Displays help on a command. For a list of commands, see !cmdinfo.");

Henzell::Config::read();

my $sigils = Henzell::Config::sigils();

# prep the new @ARGV for the help command
$ARGV[3] = 1;
$ARGV[2] =~ s/^!help\s+//i;
$ARGV[2] = "!$ARGV[2]" unless !$ARGV[2] || substr($ARGV[2], 0, 1) =~ /[\Q$sigils\E]/;

# which command are we going for?
$ARGV[2] =~ /(\S+)/;
my $requested = lc(defined($1) ? $1 : "!help");

if ($CMDPATH{$requested}) {
  print "$requested: ";
  exec $CMDPATH{$requested}, @ARGV;
} else {
  print "[[[LEARNDB: ${requested}: :::!help:${requested}:::No help for $requested (you could add help with !learn add !help:$requested <helpful text>)]]]\n"
}

File: commands/help.rb
#! /usr/bin/env ruby

require 'helper'
require 'cmd/command'
require 'cmd/user_defined_command'
require 'henzell/config'

cmdline = ARGV[2].to_s.gsub(/^\S+\s+/, '').strip

STDERR.puts("ARGV: #{ARGV.inspect}")
help("!help [CMD] displays help on a command. For a list of commands, use `!cmdinfo` for built-ins, and `!cmd -ls` for user-defined commands.",
     cmdline.empty?)

def command_help(cmdline)
  cmd_words = cmdline.split(' ')
  cmd_words[0] = Cmd::UserDefinedCommand.canonicalize_name(cmd_words[0])
  cmd = Cmd::Command.new(cmd_words.join(' '))
  case
  when cmd.builtin?
    code, output = cmd.execute(Henzell::Config.default, Helper.henzell_env, false, true)
    puts("#{cmd.command}: #{output}")
  when cmd.user_defined?
    puts "[[[LEARNDB: #{cmd.command}: :::!help:#{cmd.command}:::No help for #{cmd.command} (you could add help with !learn add !help:#{cmd.command} <helpful text>)]]]"
  else
    puts "Unknown command: #{cmd}"
  end
end

command_help(cmdline)
